--- /usr/src/kernels/linux-3.0.4/include/net/udp.h	2011-08-30 02:26:30.000000000 +0530
+++ linux-3.0.4/include/net/udp.h	2013-02-17 23:56:34.297951516 +0530
@@ -77,6 +77,25 @@
 	unsigned int		mask;
 	unsigned int		log;
 };
+
+// udp-epoll support - begin
+
+struct udp_conn_info {
+    struct udp_peer_info upi;
+    struct sock *sk;
+    struct hlist_node hnode;
+};
+
+struct udp_conn_table {
+    spinlock_t block;
+    struct hlist_head bhead;
+};
+
+#define D_UDP_CONN_SIZE 1024
+extern struct udp_conn_table udp_conn_table[D_UDP_CONN_SIZE];
+
+// udp-epoll support - end
+
 extern struct udp_table udp_table;
 extern void udp_table_init(struct udp_table *, const char *);
 static inline struct udp_hslot *udp_hashslot(struct udp_table *table,
@@ -164,8 +183,11 @@
 extern void udp_lib_unhash(struct sock *sk);
 extern void udp_lib_rehash(struct sock *sk, u16 new_hash);
 
+void udp4_lib_del_sock_upi(struct sock *sk);
+
 static inline void udp_lib_close(struct sock *sk, long timeout)
 {
+    udp4_lib_del_sock_upi(sk);
 	sk_common_release(sk);
 }
 
